### PR TITLE
Doctrine DBAL store

### PR DESCRIPTION
**Pull request by Miliooo**
I needed a key value store dbal implementation for my own project. 

The only thing I found was https://github.com/doctrine/KeyValueStore/blob/master/lib/Doctrine/KeyValueStore/Storage/DBALStorage.php and this package.

So I decided to implement one that implements this projects interface.

### CODE REVIEW

**Comment by webmozart** [comment]
Hi @Miliooo, thanks a lot for this! Could you fix the StyleCI issues please?

Did you fix the failing test you had problems with?

File: composer.json [hunk 1]
```diff
@@ -17,12 +17,14 @@
...
-        "phpunit/phpunit": "^4.6"
+        "phpunit/phpunit": "^4.6",
+        "doctrine/dbal" : "~2.4"
     },
     "suggest": {
         "predis/predis": "To enable the PredisStore",
         "basho/riak": "To enable the RiakStore",
-        "doctrine/cache": "To enable the CachedStore"
+        "doctrine/cache": "To enable the CachedStore",
+        "doctrine/dbal": "To enable the Dbalstore"
```
Comment (jdreesen) [inline composer.json]: Store with a capital s?

Comment (Miliooo) [inline composer.json]: thx fixed

**Comment by Miliooo** [comment]
Hi @webmozart, I updated the StyleCI config and also applied their patch.
Yeah I fixed the failing test, it was a logical error on my part. 

Returning null from the query, which is the same as the default value, so in my logic, the key did not exist. That's why the code now returns the database row.

I guess this is ready for review.

File: src/DbalStore.php [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+        return $result ? true : false;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function clear()
+    {
+        try {
+            $connection = $this->connection->query('DELETE FROM '.$this->table);
```
Comment (tgalopin) [inline src/DbalStore.php]: I would name this `$query`.

File: src/DbalStore.php [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+    private $connection;
+    private $table;
+
+    /**
+     * @param Connection $connection A doctrine connection instance
+     * @param string     $table      The name of the database table
+     */
+    public function __construct(Connection $connection, $table = 'store')
+    {
+        Assert::string($table, 'The table must be a string. Got: %s');
```
Comment (webmozart) [inline src/DbalStore.php]: You can use the combined constrained `stringNotEmpty()` here.

File: src/DbalStore.php [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+            $stmt = $this->connection->query('SELECT meta_key FROM '.$this->table);
+            $result = $stmt->fetchAll(\PDO::FETCH_COLUMN);
+        } catch (\Exception $e) {
+            throw ReadException::forException($e);
+        }
+
+        return $result;
+    }
+
+    public function configureTable()
```
Comment (tgalopin) [inline src/DbalStore.php]: Can you add a phpdoc comment here as it's a public method?

File: src/DbalStore.php [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+    /**
+     * {@inheritdoc}
+     */
+    public function set($key, $value)
+    {
+        KeyUtil::validate($key);
+
+        try {
+            $existing = $this->exists($key);
+        } catch (\Exception $e) {
```
Comment (webmozart) [inline src/DbalStore.php]: Can you change this to `(Exception $e)` and add `use Exception;` at the top?

File: src/DbalStore.php [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function keys()
+    {
+        try {
+            $stmt = $this->connection->query('SELECT meta_key FROM '.$this->table);
+            $result = $stmt->fetchAll(\PDO::FETCH_COLUMN);
```
Comment (webmozart) [inline src/DbalStore.php]: same here: `(PDO::...`, `use PDO;`

File: src/DbalStore.php [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+
+        return $table;
+    }
+
+    /**
+     * @param Schema $schema
+     *
+     * @return \Doctrine\DBAL\Schema\Table|null
+     */
+    public function configureSchema(Schema $schema)
```
Comment (webmozart) [inline src/DbalStore.php]: Why do we need this method?

File: src/DbalStore.php [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+
+        $table = $schema->createTable($this->table);
+
+        $table->addColumn('id', 'integer', array('autoincrement' => true));
+        $table->addColumn('meta_key', 'string', array('length' => 255));
+        $table->addColumn('meta_value', 'object');
+        $table->setPrimaryKey(array('id'));
+        $table->addUniqueIndex(array('meta_key'));
+
+        return $table;
```
Comment (webmozart) [inline src/DbalStore.php]: Why does this method return a value? How is this method used?

**Comment by webmozart** [comment]
Looks good! Can you update the README and CHANGELOG as well?

File: src/DbalStore.php (hunk 7 above)
Comment (Miliooo) [inline src/DbalStore.php]: Well I 'borrowed' that code from [broadway event store] (https://github.com/qandidate-labs/broadway/blob/master/src/Broadway/EventStore/DBALEventStore.php#L146)

I used it to setup my tests, and it makes it possible to do add a [create database command] (https://github.com/qandidate-labs/broadway/blob/master/src/Broadway/Bundle/BroadwayBundle/Command/SchemaEventStoreCreateCommand.php) in symfony. Maybe in the readme we could provide a database schema too... Or maybe add a reference to that function so users can see how they have to create their database.

File: src/DbalStore.php (hunk 8 above)
Comment (Miliooo) [inline src/DbalStore.php]: It's used for setting up my tests, and can be used to let doctrine setup your tests, see my previous comment and https://github.com/webmozart/key-value-store/pull/15/files#diff-840e95275d34986f00b5591a212cbcd2R39

File: src/DbalStore.php (hunk 4 above)
Comment (Miliooo) [inline src/DbalStore.php]: Guess I'll make it protected, since the entry point is the method configureSchema.

File: src/DbalStore.php (hunk 2 above)
Comment (Miliooo) [inline src/DbalStore.php]: I guess the right name would be statement since the method returns a statement object. Hope I'm not the only one who finds this very confusing

File: src/DbalStore.php (hunk 7 above)
Comment (webmozart) [inline src/DbalStore.php]: I see. I'd drop this method and add a public `getTableName()` then. Hiding the `hasTable()` statement here and adding a `null` check instead in the calling code seems a bit magical to me.

I'd also rename the `$table` property to `$tableName` to be clear that this is a string.

File: src/DbalStore.php (hunk 8 above)
Comment (webmozart) [inline src/DbalStore.php]: Ok :) Let's rename it to `createTable()` then?

File: src/DbalStore.php (hunk 4 above)
Comment (webmozart) [inline src/DbalStore.php]: I'd keep this as a public method, rename to `createTable()` and drop `configureSchema()` instead (as explained below).

File: src/DbalStore.php (hunk 2 above)
Comment (webmozart) [inline src/DbalStore.php]: Yes, let's call this `$stmt`.

File: src/DbalStore.php (hunk 8 above)
Comment (Miliooo) [inline src/DbalStore.php]: Well it doesn't really create a table, it's an object representation of a table. 
The schema manager already has the method `createTable.` Maybe `getTableSchema`.
This gives the following code 
```
        $tableName = self::$dbalStore->getTableName();
        if ($schema->hasTable($tableName)) {
            return;
        }

        $schemaManager->createTable(self::$dbalStore->getTableSchema());
```

I like it more then two createTableCalls on different objects. It also returns a Table object, so I like it more then getSchemaTable (which would suggest we return a schema object)

Comment (webmozart) [inline src/DbalStore.php]: I see. `getTableForCreate()` maybe?

Comment (webmozart) [inline src/DbalStore.php]:
```php
if (!$schema->hasTable(self::$dbalStore->getTableName())) {
    $schemaManager->createTable(self::$dbalStore->getTableForCreate());
}
```

Comment (Miliooo) [inline src/DbalStore.php]: cool, yeah no need to add a temporary var

**Comment by Miliooo** [comment]
Did a bunch of updates, I'll squash some commits later. If there are other changes needed let me know.

File: src/DbalStore.php [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+     */
+    public function getTableName()
+    {
+        return $this->tableName;
+    }
+
+    /**
+     * Object Representation of the table used in this class.
+     *
+     * @return \Doctrine\DBAL\Schema\Table
```
Comment (webmozart) [inline src/DbalStore.php]: You can shorten this to `@return Table` and add `use Doctrine\DBAL\Schema\Table` at the top. :)

Comment (Miliooo) [inline src/DbalStore.php]: Yes, but you have to know your abc very well because it also has to be in the right order :mortar_board: 
https://www.youtube.com/watch?v=75p-N9YKqNo

**Comment by Miliooo** [comment]
I screwed up my git commits. Not sure if you are ok with 6 different commits. I could resubmit a new pull request. Thing that got me stuck are those ```merge branch``` messages. ```git rebase -i HEAD~``` doesn't seem to work very well on those.

Been a while since I submitted pull requests, I'm just pushing to master branch all the time on my personal project.

**Comment by Miliooo** [comment]
closed in favor of https://github.com/webmozart/key-value-store/pull/17
